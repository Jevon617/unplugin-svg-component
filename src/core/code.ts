export const template = `
  <svg style="width: 1em; height: 1em; fill:currentColor ">
    <use :xlink:href="'#' + name" />
  </svg>
`
export const dts = `
declare module 'virtual:svg-component' {
  import type { PropType } from 'vue';
  const _default: import("vue").DefineComponent<{
      name: {
          type: PropType<"$svg_symbolIds">;
          default: string;
          required: true;
      };
  }, {}, unknown, {}, {}, import("vue").ComponentOptionsMixin, import("vue").ComponentOptionsMixin, {}, string, import("vue").VNodeProps & import("vue").AllowedComponentProps & import("vue").ComponentCustomProps, Readonly<import("vue").ExtractPropTypes<{
      name: {
          type: PropType<"$svg_symbolIds">;
          default: string;
          required: true;
      };
  }>>, {
      name: "$svg_symbolIds";
  }>;
  export default _default;
}
`

