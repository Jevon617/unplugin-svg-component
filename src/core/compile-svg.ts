import path from 'path'
import fs from 'fs/promises'
import fg from 'fast-glob'
import debug from 'debug'
import { optimize } from 'svgo'
import SvgCompiler from 'svg-baker'
import type { OptimizedSvg } from 'svgo'
import type { Options } from '../types'

const myDebugger = debug('unplugin-svg-component')

export default async function createSvgSprite(options: Options) {
  let svgSprite = ''
  const symbolIds = new Set()
  const { iconDir, prefix = '' } = options
  const svgCompiler = new SvgCompiler()
  const svgNames = fg.sync(['**/*.svg'], { cwd: options.iconDir })

  for (const svgName of svgNames) {
    const svgPath = path.resolve(iconDir, svgName)
    const svgContent = await fs.readFile(svgPath)
    const symbolId = createSymbolId(svgName, prefix)

    let { data: OptimizedSvgContent } = await optimize(svgContent) as OptimizedSvg

    OptimizedSvgContent = OptimizedSvgContent
      .replace(/stroke="[a-zA-Z#0-9]*"/g, 'stroke="currentColor"')
      .replace(/fill="[a-zA-Z#0-9]*"/g, (p: string) => {
        if (p.includes('none'))
          return p
        else
          return 'fill="currentColor"'
      })

    const svgSymbol = await svgCompiler.addSymbol({
      path: svgPath,
      content: OptimizedSvgContent,
      id: symbolId,
    })

    symbolIds.add(symbolId)
    svgSprite += `${svgSymbol.render() || ''}`
  }

  myDebugger(svgSprite)
  myDebugger(symbolIds)
  return {
    symbolIds,
    svgSprite,
  }
}

function createSymbolId(svgName: string, prefix: string) {
  const nameArr = svgName.split('/')
  if (prefix)
    nameArr.unshift(prefix)
  return nameArr.join('-').replace('.svg', '')
}
