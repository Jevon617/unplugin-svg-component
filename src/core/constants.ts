export const LOAD_EVENT = 'svg-component:reload'
export const UPDATE_EVENT = 'svg-component:update'
export const MODULE_NAME = 'virtual:svg-component'
