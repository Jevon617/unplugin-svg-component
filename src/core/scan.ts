import fs from 'node:fs/promises'
import path from 'node:path'
import fg from 'fast-glob'
import type { Options } from '../types'

export default async function scanUsedSvgNames(options: Options) {
  const { componentName, scanGlob, iconDir, scanStrategy, symbolIdFormatter, prefix } = options

  const allSvgNames = fg.sync(['**/*.svg'], { cwd: iconDir })
    .map(n => symbolIdFormatter!(n, prefix!))
    .sort((a, b) => b.length - a.length)

  const svgNameRE = new RegExp(`(${allSvgNames.join('|')})`, 'gm')

  const svgCompnentRE = new RegExp(
    `\\<\\s*${componentName}[^\\<]+name=[\\"\\'](\\S+)[\\"\\'][^\\<]*\\/\\>`, 'g',
  )
  const filenames = await fg(scanGlob!, {
    onlyFiles: true,
    ignore: [
      '**/node_modules',
      '**/dist',
      '**/*.d.ts',
      'vite.config.ts',
      'vite.config.js',
    ],
  })

  const scanExecutors = filenames.map(async (f) => {
    const code = await fs.readFile(path.resolve(process.cwd(), f), 'utf-8')

    if (scanStrategy === 'component')
      return Array.from(code.matchAll(svgCompnentRE)).map(match => match[1])
    else
      return [...(code.match(svgNameRE) || [])]
  })

  const svgNames = await Promise.all(scanExecutors)
  return svgNames.flat()
}
