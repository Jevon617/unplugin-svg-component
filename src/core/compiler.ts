import path from 'path'
import fs from 'fs/promises'
// import { compileTemplate } from '@vue/compiler-sfc'
import { getPackageInfo, importModule } from 'local-pkg'
import type { Options } from '../types'
import { dts, golbalDts, template } from './code'
import { replace } from './utils'

export async function compileComponent(componentName: string) {
  let code = ''
  const vueVerison = await getVueVersion()

  if (vueVerison === 'vue2') {
    const { compile } = await importModule('vue-template-compiler')
    const transpile = (await importModule('vue-template-es2015-compiler')).default

    const { render } = compile(template)
    const toFunction = (code: string): string => {
      return `function () {${code}}`
    }
    const res = transpile(`var __render__ = ${toFunction(render as any)}\n`, {})
    code = res.replace(/\s__(render|staticRenderFns)__\s/g, ' $1 ')
  }
  else {
    const compilerSfc = await importModule('@vue/compiler-sfc')
    const res = compilerSfc.compileTemplate({
      source: template,
      id: '__svg-icon__',
      filename: 'svg-icon.vue',
    })
    code = res.code.replace(/export/g, '')
  }

  code += `
    \nexport default{
      name: "${componentName}",
      props: {
        name: {
          type: String,
          required: true
        }
      },
      render
    }
  `
  return code
}

async function getVueVersion() {
  try {
    const result = await getPackageInfo('vue')
    if (!result)
      return null
    return result.version.startsWith('2.') ? 'vue2' : 'vue3'
  }
  catch {
    return null
  }
}

export function genDts(symbolIds: Set<string>, options: Options) {
  fs.writeFile(
    path.resolve(options.dtsDir!, './svg-component.d.ts'),
    replace(dts, symbolIds, options.componentName!),
  )
  fs.writeFile(
    path.resolve(options.dtsDir!, './svg-component-global.d.ts'),
    replace(golbalDts, symbolIds, options.componentName!),
  )
}
