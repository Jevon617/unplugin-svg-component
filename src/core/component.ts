import path from 'path'
import fs from 'fs/promises'
import { compileTemplate } from '@vue/compiler-sfc'
import type { Options } from '../types'
import { dts, golbalDts, template } from './code'
import { replace } from './utils'
export function compileComponent(componentName: string) {
  let { code } = compileTemplate({
    source: template,
    id: '__svg-icon__',
    filename: 'svg-icon.vue',
  })

  code = code.replace(/export/g, '')
  code += `
    \nexport default{
      name: "${componentName}",
      props: {
        name: {
          type: String,
          required: true
        }
      },
      render
    }
  `
  return code
}

export function genDts(symbolIds: Set<string>, options: Options) {
  fs.writeFile(
    path.resolve(options.dtsDir!, './svg-icon.d.ts'),
    replace(dts, symbolIds, options.componentName!),
  )
  fs.writeFile(
    path.resolve(options.dtsDir!, './svg-icon-global.d.ts'),
    replace(golbalDts, symbolIds, options.componentName!),
  )
}
