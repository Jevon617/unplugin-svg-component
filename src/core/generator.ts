import path from 'path'
import fs from 'fs/promises'
import { getPackageInfo, importModule, isPackageExists } from 'local-pkg'
import type { Options, VueVersion } from '../types'
import { dts, golbalDts, reactDts, reactTemplate, template } from './snippets'
import { replace, transformStyleStrToObject } from './utils'
import createSvgSprite from './sprite'
import { LOAD_EVENT, UPDATE_EVENT, XMLNS, XMLNS_LINK } from './constants'

export async function genModuleCode(options: Options, usedSvgNames: string[] | string, hmr: boolean) {
  const isVueProject = detectProjectType(options)
  const component = await genComponent(options, isVueProject)
  const svgSpriteDomId = options.svgSpriteDomId

  const { symbolIds, symbols, symbolCache } = await createSvgSprite(options, usedSvgNames)
  const xmlns = `xmlns="${XMLNS}"`
  const xmlnsLink = `xmlns:xlink="${XMLNS_LINK}"`
  const symbolHtml = Array.from(symbols).join('')
    .replace(new RegExp(xmlns, 'g'), '')
    .replace(new RegExp(xmlnsLink, 'g'), '')

  // only generate dts in serve
  if (options?.dts && !Array.isArray(usedSvgNames))
    genDts(symbolIds, options, isVueProject)

  const hmrCode = `
    if (import.meta.hot) {
      import.meta.hot.on("${LOAD_EVENT}", ({svgSymbolHtml}) => {
        svgDom.innerHTML = svgSymbolHtml
      })

      import.meta.hot.on("${UPDATE_EVENT}", ({symbolId, newSvgSymbol}) => {
        var oldSymbolDom = svgDom.querySelector('#' + symbolId)
        var tempDom = document.createElementNS('${XMLNS}', 'svg');
        tempDom.innerHTML = newSvgSymbol
        var newSymbolDom = tempDom.children[0]
        svgDom.replaceChild(newSymbolDom, oldSymbolDom)
      })
    }
  `
  const code = `
    ${component}

    var svgDom = document.querySelector('#${svgSpriteDomId}') 
      || document.createElementNS('${XMLNS}', 'svg');

    ${hmr ? hmrCode : ''}
    if (typeof window !== 'undefined') {
      function loadSvgs() {
        var body = document.body;
        svgDom.style.position = 'absolute';
        svgDom.style.width = '0';
        svgDom.style.height = '0';
        svgDom.id = '${svgSpriteDomId}';
        svgDom.setAttribute('xmlns','${XMLNS}');
        svgDom.setAttribute('xmlns:link','${XMLNS_LINK}');
        svgDom.innerHTML = ${JSON.stringify(symbolHtml)};
        body.append(svgDom)
      }
      if(document.readyState === 'loading') {
        document.addEventListener('DOMContentLoaded', loadSvgs);
      } else {
        loadSvgs()
      }
    }
   `
  return {
    symbolCache,
    symbolIds,
    symbols,
    code,
  }
}

export function genDts(symbolIds: Set<string>, options: Options, isVueProject: boolean) {
  if (isVueProject) {
    fs.writeFile(
      path.resolve(options.dtsDir!, './svg-component.d.ts'),
      replace(dts, symbolIds, options.componentName!),
    )
    fs.writeFile(
      path.resolve(options.dtsDir!, './svg-component-global.d.ts'),
      replace(golbalDts, symbolIds, options.componentName!),
    )
  }
  else {
    fs.writeFile(
      path.resolve(options.dtsDir!, './svg-component.d.ts'),
      replace(reactDts, symbolIds, options.componentName!),
    )
  }
}

async function genComponent(options: Options, isVueProject: boolean) {
  const { componentStyle, componentName, vueVersion } = options
  if (!isVueProject) {
    return reactTemplate.replace(/\$component_name/, componentName!)
      .replace(/\$component_style/, JSON.stringify(transformStyleStrToObject(componentStyle!)))
  }

  const vueVerison = await getVueVersion(vueVersion!)
  if (!vueVerison)
    return 'export default {}'

  const compilers = {
    vue2: compileVue2Template,
    vue3: compileVue3Template,
  }

  const processedTemplate = template.replace(/\$component_style/, componentStyle!)

  let code = await compilers[vueVerison](processedTemplate)

  code += `
    \nexport default{
      name: "${componentName}",
      props: {
        name: {
          type: String,
          required: true
        }
      },
      render
    }
  `
  return code
}

async function compileVue3Template(template: string): Promise<string> {
  const { compileTemplate } = await importModule('@vue/compiler-sfc')
  const { code } = compileTemplate({
    source: template,
    id: '__svg-component__',
    filename: 'virtual:svg-component.vue',
  })
  return code.replace(/export/g, '')
}

async function compileVue2Template(template: string): Promise<string> {
  const { compile } = await importModule('vue-template-compiler')
  const transpile = (await importModule('vue-template-es2015-compiler')).default
  const { render } = compile(template)
  const toFunction = (code: string): string => {
    return `function () {${code}}`
  }
  const res = transpile(`var __render__ = ${toFunction(render as any)}\n`, {})
  const code = res.replace(/\s__(render|staticRenderFns)__\s/g, ' $1 ')

  return code
}

async function getVueVersion(vueVerison: VueVersion): Promise<'vue2' | 'vue3' | null > {
  if (vueVerison === 'auto') {
    try {
      const result = await getPackageInfo('vue')
      if (!result)
        return null
      return result.version.startsWith('2.') ? 'vue2' : 'vue3'
    }
    catch {
      return null
    }
  }
  else {
    return `vue${vueVerison}`
  }
}

function detectProjectType(options: Options) {
  if (options.projectType === 'auto') {
    try {
      if (isPackageExists('vue'))
        return true
      else if (isPackageExists('react'))
        return false
      else
        throw new Error('[unpluign-svg-component] can\'t detect your project type, please set options.projectType.')
    }
    catch {
      throw new Error('[unpluign-svg-component] can\'t detect your project type, please set options.projectType.')
    }
  }
  else {
    return options.projectType === 'vue'
  }
}
