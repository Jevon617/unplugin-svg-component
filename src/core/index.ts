import { createUnplugin } from 'unplugin'
import type { ViteDevServer } from 'vite'
import type { Options } from '../types'
import watchIconDir from './watcher'
import { compileComponent, genDts } from './compiler'
import createSvgSprite, { svgSymbols, symbolIds } from './sprite'
import { LOAD_EVENT, MODULE_NAME, UPDATE_EVENT } from './constants'

let virtualModuleCode = ''
let TransformPluginContext

function resolveOptions(options: Options): Options {
  const defaultOptions = {
    componentName: 'SvgIcon',
    dtsDir: process.cwd(),
    svgSpriteDomId: '__svg__icons__dom__',
  }
  return {
    ...defaultOptions,
    ...options,
  }
}

export default createUnplugin<Options | undefined>(options => ({
  name: 'unplugin-svg-component',
  async buildStart() {
    options = resolveOptions(options!)
    virtualModuleCode = await createCode(options, false)
  },
  resolveId(id: string) {
    if (id === MODULE_NAME)
      return id
  },
  loadInclude(id) {
    return id === MODULE_NAME
  },
  async load() {
    return {
      code: virtualModuleCode,
    }
  },
  vite: {
    transform(this) {
      // eslint-disable-next-line @typescript-eslint/no-this-alias
      TransformPluginContext = this
    },
    configureServer(server: ViteDevServer) {
      watchIconDir(options!, server)
      server.middlewares.use(async (req, res, next) => {
        if (req.url === `/@id/${MODULE_NAME}`) {
          const code = await createCode(options!, true)

          const importAnalysisTransform = server.config.plugins.find(
            plugin => plugin.name === 'vite:import-analysis',
          )?.transform as any

          const transformResult = await importAnalysisTransform.apply(
            TransformPluginContext, [code, MODULE_NAME, { ssr: false }],
          )
          res.statusCode = 200
          res.setHeader('Content-Type', 'application/javascript')
          res.setHeader('Cache-Control', 'no-cache')
          res.end(transformResult.code)
        }
        else {
          next()
        }
      })
    },
  },
}))

async function createCode(options: Options, hmr: boolean) {
  const componentCode = await compileComponent(options.componentName!)
  await createSvgSprite(options)
  const svgSpriteDomId = options.svgSpriteDomId || '__svg__icons__dom__'

  if (options?.dts)
    genDts(symbolIds, options)

  const hmrCode = `
    if (import.meta.hot) {
      import.meta.hot.on("${LOAD_EVENT}", ({svgSymbolHtml}) => {
        svgDom.innerHTML = svgSymbolHtml
      })

      import.meta.hot.on("${UPDATE_EVENT}", ({symbolId, newSvgSymbol}) => {
        var oldSymbolDom = svgDom.querySelector('#' + symbolId)
        var tempDom = document.createElementNS('http://www.w3.org/2000/svg', 'svg');
        tempDom.innerHTML = newSvgSymbol
        var newSymbolDom = tempDom.children[0]
        svgDom.replaceChild(newSymbolDom, oldSymbolDom)
      })
    }
  `

  const code = `
    ${componentCode}

    var svgDom = document.querySelector('#${svgSpriteDomId}') 
      || document.createElementNS('http://www.w3.org/2000/svg', 'svg');

    ${hmr ? hmrCode : ''}
    if (typeof window !== 'undefined') {
      function loadSvgs() {
        var body = document.body;
        svgDom.style.position = 'absolute';
        svgDom.style.width = '0';
        svgDom.style.height = '0';
        svgDom.id = '${svgSpriteDomId}';
        svgDom.setAttribute('xmlns','http://www.w3.org/2000/svg');
        svgDom.setAttribute('xmlns:link','http://www.w3.org/1999/xlink');
        svgDom.innerHTML = ${JSON.stringify(Array.from(svgSymbols).join('\n'))};
        body.append(svgDom)
      }
      if(document.readyState === 'loading') {
        document.addEventListener('DOMContentLoaded', loadSvgs);
      } else {
        loadSvgs()
      }
    }
   `
  return code
}
