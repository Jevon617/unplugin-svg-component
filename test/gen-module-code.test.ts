import path from 'path'
import { expect, test } from 'vitest'
import { genModuleCode } from '../src/core/generator'

test('createSprite', async () => {
  const { code } = await genModuleCode({
    iconDir: path.resolve(__dirname, './icons'),
  }, true)
  expect(code).matchSnapshot()
})
