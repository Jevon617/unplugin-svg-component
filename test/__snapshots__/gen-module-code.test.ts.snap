// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`createSprite 1`] = `
"
    import { createElementVNode as _createElementVNode, openBlock as _openBlock, createElementBlock as _createElementBlock } from \\"vue\\"

const _hoisted_1 = {
  xmlns: \\"http://www.w3.org/2000/svg\\",
  \\"xmlns:xlink\\": \\"http://www.w3.org/1999/xlink\\",
  style: {}
}
const _hoisted_2 = [\\"xlink:href\\"]

 function render(_ctx, _cache) {
  return (_openBlock(), _createElementBlock(\\"svg\\", _hoisted_1, [
    _createElementVNode(\\"use\\", {
      \\"xlink:href\\": '#' + _ctx.name
    }, null, 8 /* PROPS */, _hoisted_2)
  ]))
}

export default{
  name: \\"SvgIcon\\",
  props: {
    name: {
      type: String,
      required: true
    }
  },
  render
}
    
export const svgNames = [\\"icon-addUser\\",\\"icon-card2\\",\\"logo\\",\\"common-icon-add\\",\\"common-icon-add3\\",\\"common-icon-addUser\\",\\"common-icon-addUsers\\",\\"common-icon-addx\\",\\"common-icon-apple\\",\\"common-icon-banner\\",\\"common-logo\\"]

    
if (import.meta.hot) {
  var svgDom = document.querySelector('#sprite-id') 
  import.meta.hot.on(\\"svg-component:reload\\", ({svgSymbolHtml}) => {
    svgDom.innerHTML = svgSymbolHtml
  })

  import.meta.hot.on(\\"svg-component:update\\", ({symbolId, newSvgSymbol}) => {
    var oldSymbolDom = svgDom.querySelector('#' + symbolId)
    var tempDom = document.createElementNS('http://www.w3.org/2000/svg', 'svg');
    tempDom.innerHTML = newSvgSymbol
    var newSymbolDom = tempDom.children[0]
    svgDom.replaceChild(newSymbolDom, oldSymbolDom)
  })
}

   "
`;
